internal/app/machined/pkg/controllers/network/hostdns_config.go

```diff
@@ -0,0 +1,202 @@
+	Cmdline *procfs.Cmdline
+}
+
+// Name implements controller.Controller interface.
+func (ctrl *HostDNSConfigController) Name() string {
+	return "network.HostDNSConfigController"
+}
+
+// Inputs implements controller.Controller interface.
+func (ctrl *HostDNSConfigController) Inputs() []controller.Input {
+	return []controller.Input{
+		{
+			Namespace: config.NamespaceName,
+			Type:      config.MachineConfigType,
+			ID:        optional.Some(config.V1Alpha1ID),
+			Kind:      controller.InputWeak,
+		},
+	}
+}
+
+// Outputs implements controller.Controller interface.
+func (ctrl *HostDNSConfigController) Outputs() []controller.Output {
+	return []controller.Output{
+		{
+			Type: network.HostDNSConfigType,
+			Kind: controller.OutputExclusive,
+		},
+		{
+			Type: network.AddressSpecType,
+			Kind: controller.OutputShared,
+		},
+	}
+}
+
+// Run implements controller.Controller interface.
+//
+//nolint:gocyclo
+func (ctrl *HostDNSConfigController) Run(ctx context.Context, r controller.Runtime, logger *zap.Logger) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-r.EventCh():
+		}
+
+		var cfgProvider talosconfig.Config
+
+		cfg, err := safe.ReaderGetByID[*config.MachineConfig](ctx, r, config.V1Alpha1ID)
+		if err != nil {
+			if !state.IsNotFoundError(err) {
+				return fmt.Errorf("error getting config: %w", err)
+			}
+		} else if cfg.Config().Machine() != nil {
+			cfgProvider = cfg.Config()
+		}
+
+		var oldServiceAddr, newServiceAddr netip.Addr
+
+		if err := safe.WriterModify(ctx, r, network.NewHostDNSConfig(network.HostDNSConfigID), func(res *network.HostDNSConfig) error {
+			res.TypedSpec().ListenAddresses = []netip.AddrPort{
+				netip.MustParseAddrPort("127.0.0.53:53"),
+				netip.MustParseAddrPort("[::1]:53"),
+			}
+
+			oldServiceAddr = res.TypedSpec().ServiceHostDNSAddress
+			res.TypedSpec().ServiceHostDNSAddress = netip.Addr{}
+
+			if cfgProvider == nil {
+				res.TypedSpec().Enabled = false
+
+				return nil
+			}
+
+			res.TypedSpec().Enabled = cfgProvider.Machine().Features().HostDNS().Enabled()
+
+			if cfgProvider.Machine().Features().HostDNS().ForwardKubeDNSToHost() {
+				serviceCIDRStr := cfgProvider.Cluster().Network().ServiceCIDRs()[0]
+
+				serviceCIDR, err := netip.ParsePrefix(serviceCIDRStr)
+				if err != nil {
+					return fmt.Errorf("error parsing service CIDR: %w", err)
+				}
+
+				newServiceAddr = serviceCIDR.Addr()
+				for range 9 {
+					newServiceAddr = newServiceAddr.Next()
+				}
+
+				res.TypedSpec().ListenAddresses = append(res.TypedSpec().ListenAddresses, netip.AddrPortFrom(newServiceAddr, 53))
+				res.TypedSpec().ServiceHostDNSAddress = newServiceAddr
+			}
+
+			return nil
+		}); err != nil {
+			return fmt.Errorf("error writing host dns config: %w", err)
+		}
+
+		if oldServiceAddr == newServiceAddr {
+			continue
+		}
+
+		if !value.IsZero(oldServiceAddr) {
+			// new and old service addresses are different, so we need to destroy the old address spec
```

Comment:
this is a bit harder to manage than it should be, we can simplify it to dropping all addresspec resources we haven't touched:  https://github.com/siderolabs/talos/blob/5c0f74b377757ffee8f455ac2bcd9d81e3bf7717/internal/app/machined/pkg/controllers/siderolink/manager.go#L456-L482

Reply:
Fixed!